File: Algorithm/Python/503.next-greater-element-ii.py
#
# @lc app=leetcode id=503 lang=python
#
# [503] Next Greater Element II
#

# @lc code=start
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        result = [-1] * len(nums)
        stack = []
        for _ in range(2):
            for i, num in enumerate(nums):
                while stack and nums[stack[-1]] < num:
                    idx = stack.pop()
                    if result[idx] == -1:
                        result[idx] = num
                stack.append(i)
        return result
        
# @lc code=end

File: Algorithm/C++/503.next-greater-element-ii.cpp
/*
 * @lc app=leetcode id=503 lang=cpp
 *
 * [503] Next Greater Element II
 */

// @lc code=start

#include <vector>
#include <stack>

class Solution {
public:
    std::vector<int> nextGreaterElements(std::vector<int>& nums) {
        std::vector<int> result(nums.size(), -1);
        std::stack<int> s;        
        for (int _iter=0; _iter < 2; _iter++) {
            for (int i = 0; i < nums.size(); i++) {
                while (!s.empty() && nums[s.top()] < nums[i]) {
                    if (result[s.top()] == -1) {
                        result[s.top()] = nums[i];
                    }
                    s.pop();
                }
                s.push(i);
            }
        }
        return result;
    }
};
// @lc code=end

